emasamplerate.go

```diff
@@ -0,0 +1,297 @@
+package dynsampler
+
+import (
+	"math"
+	"sort"
+	"sync"
+	"time"
+)
+
+// EMASampleRate implements Sampler and attempts to average a given sample rate,
+// weighting rare traffic and frequent traffic differently so as to end up with
+// the correct average. This method breaks down when total traffic is low
+// because it will be excessively sampled.
+//
+// Based on the AvgSampleRate implementation, EMASampleRate differs in that rather
+// than compute rate based on a periodic sample of traffic, it maintains an Exponential
+// Moving Average of counts seen per key, and adjusts this average at regular intervals.
+// The weight applied to more recent intervals is defined by `weight`, a number between
+// (0, 1) - larger values weight the average more toward recent observations.
```

Comment:
maybe add a rephrase as well - "weight the average more towards recent observations - in other words, the effect of a change in rate decays more rapidly" or something?